contracts/external/traders/CanonicalOrders.sol

```diff
@@ -0,0 +1,842 @@
+                FILE,
+                "Order invalid signature",
+                orderInfo.orderHash
+            );
+        } else {
+            Require.that(
+                orderStatus != OrderStatus.Canceled,
+                FILE,
+                "Order canceled",
+                orderInfo.orderHash
+            );
+            assert(orderStatus == OrderStatus.Approved);
+        }
+
+        // verify price
+        TradeArgs memory tradeArgs = orderInfo.tradeArgs;
+        bool validPrice = isBuy(orderInfo.order)
+            ? tradeArgs.price <= orderInfo.order.limitPrice
+            : tradeArgs.price >= orderInfo.order.limitPrice;
+        Require.that(
+            validPrice,
+            FILE,
+            "Fill invalid price"
+        );
+
+        // verify fee
+        bool validFee = isNegativeFee(orderInfo.order)
+            ? (tradeArgs.fee >= orderInfo.order.limitFee) && tradeArgs.isNegativeFee
+            : (tradeArgs.fee <= orderInfo.order.limitFee) || tradeArgs.isNegativeFee;
+        Require.that(
+            validFee,
+            FILE,
+            "Fill invalid fee"
+        );
+
+        return orderInfo;
+    }
+
+    // ============ Private Parsing Functions ============
+
+    /**
+     * Returns the EIP712 hash of an order.
+     */
+    function getOrderHash(
+        Order memory order
+    )
+        private
+        view
+        returns (bytes32)
+    {
+        // compute the overall signed struct hash
+        /* solium-disable-next-line indentation */
+        bytes32 structHash = keccak256(abi.encode(
+            EIP712_ORDER_STRUCT_SCHEMA_HASH,
+            order
+        ));
+
+        // compute eip712 compliant hash
+        /* solium-disable-next-line indentation */
+        return keccak256(abi.encodePacked(
+            EIP191_HEADER,
+            EIP712_DOMAIN_HASH,
+            structHash
+        ));
+    }
+
+    /**
+     * Parses out a signature from call data.
+     */
+    function parseSignature(
+        bytes memory data
+    )
+        private
+        pure
+        returns (bytes memory)
+    {
+        Require.that(
+            data.length == NUM_ORDER_BYTES + NUM_SIGNATURE_BYTES,
+            FILE,
+            "Cannot parse signature from data"
+        );
+
+        bytes memory signature = new bytes(NUM_SIGNATURE_BYTES);
+
+        uint256 sigOffset = NUM_ORDER_BYTES;
+        /* solium-disable-next-line security/no-inline-assembly */
+        assembly {
+            let sigStart := add(data, sigOffset)
+            mstore(add(signature, 0x020), mload(add(sigStart, 0x20)))
+            mstore(add(signature, 0x040), mload(add(sigStart, 0x40)))
+            mstore(add(signature, 0x042), mload(add(sigStart, 0x42)))
+        }
+
+        return signature;
+    }
+
+    function isBuy(
+        Order memory order
+    )
+        private
+        pure
+        returns (bool)
+    {
+        return (order.flags & IS_BUY_FLAG) != bytes32(0);
```

Comment:
it's a bit weird that any non-zero value in the byte that matches the flag will evaluate to true. So you could have orders with different values in that flag field which could mean the same thing

Reply:
As it's also the salt field, that's intended, although I can change the bitmask to be a single bit instead

Reply:
Hm ok, I guess it's fine as is

Reply:
Changed to the last 3 bits